### ISSUE TITLE
Pallas GPU FlashAttention forward only works when sequence lengths are multiples of block size

### ISSUE THREAD

**Issue by yuzu-ido** [jax-ml/jax#27224]
### Description

Hi Jax team! Thank you for all your hard work. I am interested in using FlashAttention with the Pallas (Triton backend) implementation on GPU. After some testing, it seems that the forward pass numerical output is only close to the reference if the sequence lengths are multiples of block size. Would it be possible to add functionality to support any sequence length or at least a warning specifying this so users know? (For my use case, I can work with this by changing the padding or block size.) Thank you for your consideration! I ran my tests on Nvidia A10G. Here is some reproducible code:
```
import jax
import jax.numpy as jnp
from jax.experimental.pallas.ops.gpu.attention import BlockSizes, mha, mha_reference

head_dim = 16
key = jax.random.key(100)
num_heads = 8
q_seq = 192
kv_seq = 192
block_size = 128
keys = jax.random.split(key, 3)
query = jax.random.normal(keys[0], (1, q_seq, num_heads, head_dim), dtype=jnp.bfloat16)
key = jax.random.normal(keys[1], (1, kv_seq, num_heads, head_dim), dtype=jnp.bfloat16)
value = jax.random.normal(keys[2], (1, kv_seq, num_heads, head_dim), dtype=jnp.bfloat16)
attn_flash = mha(
    query,
    key,
    value,
    None,
    block_sizes=BlockSizes(
        block_q=block_size,
        block_k=block_size,
        block_q_dkv=block_size,
        block_kv_dkv=block_size,
        block_q_dq=block_size,
        block_kv_dq=block_size,
    ),
)
attn_ref = mha_reference(query, key, value, None)
print(f"no flash vs flash {abs(attn_flash - attn_ref).max()}")
```
Output:
```
no flash vs flash 2.4726288318634033
```
If instead, block_size = 64 (= 192/3): 
```
no flash vs flash 0.00853872299194336
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.5.2
jaxlib: 0.5.1
numpy:  2.0.0
python: 3.10.9 (main, Jan 16 2023, 22:37:31) [Clang 15.0.7 ]
device info: NVIDIA A10G-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='crepe', release='6.8.0-1021-aws', version='#23~22.04.1-Ubuntu SMP Tue Dec 10 16:50:46 UTC 2024', machine='x86_64')


$ nvidia-smi
Tue Mar 18 18:43:35 2025       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.06             Driver Version: 535.183.06   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A10G                    On  | 00000000:00:1E.0 Off |                    0 |
|  0%   28C    P0              59W / 300W |  17427MiB / 23028MiB |      2%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A   1935989      C   ...86_64-unknown-linux-gnu/bin/python3    17160MiB |
|    0   N/A  N/A   2098227      C   ...86_64-unknown-linux-gnu/bin/python3      254MiB |
+---------------------------------------------------------------------------------------+
```

**Comment by rdyro** [comment]
Thanks for reporting this!

I'm trying to add checks on sequence length being divisible by block size. The current recommended workaround will be to pad the sequences to a number divisible by the block sizes - but the kernel should throw an intelligible error if this isn't the case instead of returning numerically incorrect results.
